mysql_tests/test_engine.py

```diff
@@ -0,0 +1,403 @@
+    loop.create_task(second())
+    t3 = loop.create_task(third())
+    assert await t1 is await t3
+    await e.close()
+
+
+# noinspection PyProtectedMember
+async def test_lazy(mocker):
+    engine = await create_engine(MYSQL_URL, minsize=1, maxsize=1)
+    init_size = qsize(engine)
+    async with engine.acquire(lazy=True):
+        assert qsize(engine) == init_size
+        assert len(engine._ctx.get()) == 1
+    assert engine._ctx.get() is None
+    assert qsize(engine) == init_size
+    async with engine.acquire(lazy=True):
+        assert qsize(engine) == init_size
+        assert len(engine._ctx.get()) == 1
+        assert await engine.scalar("select 1")
+        assert qsize(engine) == init_size - 1
+        assert len(engine._ctx.get()) == 1
+    assert engine._ctx.get() is None
+    assert qsize(engine) == init_size
+
+    loop = asyncio.get_event_loop()
+    fut = loop.create_future()
+
+    async def block():
+        async with engine.acquire():
+            fut.set_result(None)
+            await asyncio.sleep(0.3)
+
+    blocker = loop.create_task(block())
+    await fut
+    init_size_2 = qsize(engine)
+    ctx = engine.acquire(lazy=True)
+    conn = await ctx.__aenter__()
+    t1 = loop.create_task(conn.execution_options(timeout=0.1).scalar("select 1"))
+    t2 = loop.create_task(ctx.__aexit__(None, None, None))
+    with pytest.raises(asyncio.TimeoutError):
+        await t1
+    assert not await t2
+    assert qsize(engine) == init_size_2
+    await blocker
+    assert qsize(engine) == init_size
+
+    fut = loop.create_future()
+    blocker = loop.create_task(block())
+    await fut
+    init_size_2 = qsize(engine)
+
+    async def acquire_failed(*args, **kwargs):
+        await asyncio.sleep(0.1)
+        raise ValueError()
+
+    mocker.patch("aiomysql.pool.Pool.acquire", new=acquire_failed)
+    ctx = engine.acquire(lazy=True)
+    conn = await ctx.__aenter__()
+    t1 = loop.create_task(conn.scalar("select 1"))
+    t2 = loop.create_task(conn.release(permanent=False))
+    with pytest.raises(ValueError):
+        await t1
+    assert not await t2
+    assert qsize(engine) == init_size_2
+
+    await conn.release(permanent=False)
+    assert qsize(engine) == init_size_2
+
+    await blocker
+    assert qsize(engine) == init_size
+    await engine.close()
+
+
+async def test_release(engine):
+    init_size = qsize(engine)
+    async with engine.acquire() as conn:
+        assert await conn.scalar("select 8") == 8
+        await conn.release(permanent=False)
+        assert await conn.scalar("select 8") == 8
+        await conn.release(permanent=False)
+    with pytest.raises(ValueError, match="released permanently"):
+        await conn.scalar("select 8")
+    with pytest.raises(ValueError, match="already released"):
+        await conn.release()
+
+    conn = await engine.acquire()
+    assert await conn.scalar("select 8") == 8
+    await conn.release(permanent=False)
+    assert await conn.scalar("select 8") == 8
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5587950)